### ISSUE TITLE
RFC Defined Input/Outputs

### ISSUE THREAD

**Issue by aaronmallen** [aaronmallen/activeinteractor#263]
#### Elevator pitch, describe and sell us your feature request

**As a** a developer,
**I want** To be able to define what inputs a given interactor expects and what outputs are expected to be returned
**so that** I can enforce, validate, and document expected arguments for an interactor and what it should return.

#### Is your feature request related to a problem

Hash arguments become hard to parse and maintain documentation for as an application scales.  Explicitly defining what arguments an interactor expects and what it returns as part of the gems DSL will greatly increase scalability when using the ActiveInteractor gem.
 
#### Have you considered any alternatives

You could simply ensure hash arguments for an interactor are always documented via `rdoc` or `yard` but inline documentation is difficult to maintain and impossible to enforce.
 
#### Additional Comments

The [graphql gem](https://github.com/rmosolgo/graphql-ruby)'s `field` method for a query or mutation could be a really good way to implement this feature. The ideal resolution could look something like this:

```ruby
class AuthenticateUser < ActiveInteractor::Base
  input :email, String, 'An email address', required: true
  input :password, String, 'A password', required: true
  
  output :user, User, 'A user', required: true
  output :token, String, 'An auth token', required: true
  
  input_validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }

  def perform
    context.user = User.authenticate(
      context.email,
      context.password
    )
    context.token = context.user.secret_token
  end
end

context = AuthenticateUser.perform(email: 'hello@aaronmallen.me', password: 'password')
context.successful?
#=> true
context.user 
#=> <#User ...>

context = AuthenticateUser.perform(email: 1, password: 'password')
context.successful?
#=> false
context.errors[:email]
#=> ["must be type of String"]

context = AuthenticateUser.perform(email: nil, password: 'password')
context.successful?
#=> false
context.errors[:email]
#=> ["can't be blank"]

context = AuthenticateUser.perform(email: 'hello@@aaronmallen.me', password: 'password')
context.successful?
#=> false
context.errors[:email]
#=> ["is invalid"]

User.find_by(email: 'hello@aaronmallen.me')
#=> nil
context = AuthenticateUser.perform(email: 'hello@aaronmallen.me', password: 'password')
context.successful?
#=> false
context.errors[:user]
#=> ["can't be blank"]
```
### Issues in this epic

**Comment by jaysignorello** [comment]
@aaronmallen from the above description, you're trying to make the interactor have some features of a static typed language. I can certainly understand how that would be helpful to make a larger project scalable and reduce bugs.

I like the idea of using the graphql gem dsl, but I think the implementation could use some refinements.

As of right now, you're intermixing argument errors with data validation rules, which imo isn't great. In a static typed language, if we passed the wrong type to an API, we'd get errors from the compiler. Since that's not possible with Ruby, the standard way to deal with a similar situation in Ruby would be to throw an exception (i.e. `ArgumentError`). The earlier an engineer can realize they made a mistake in development the better.

In terms of `output`, I really like how graphql gem handles it. They define the output fields similar to how you've designed yours, however the way they have the developer return them is by just returning a single value or hash. More functional approach.

```
...
  def perform
    user = User.authenticate(
      context.email,
      context.password
    )

    {user: user, token: user.secret_token}
  end
...
```

If you're looking for more of a OOP pattern, I'd suggest assigning to a different instance variable. `context` should be reserved for the context of the original request imo.

```
...
  def perform
    output.user = User.authenticate(
      context.email,
      context.password
    )
    output.token = output.user.secret_token
  end
...
```
